### ISSUE TITLE
BulkUpdate doesn't support parameters

### ISSUE THREAD

**Issue by ducman** [MoonStorm/Dapper.FastCRUD#61]
After reading through the code, I found that `BulkUpdate(...)` doesn't build the query using the provided parameters via `statementOptions`. The actual entity object is used as the parameters. This put a lot of limitation to do bulk update.

Is there anyway to append the parameters from 'statementOptions' to the entity object? Of course you can't just literal just do that but maybe the use of `DynamicParameters` with `TypeDescriptor` to build out a new parameters object to use in the query?

**Comment by MoonStorm** [comment]
The bulk update is meant to be used together with  the [entity mapping override](https://github.com/MoonStorm/Dapper.FastCRUD/wiki/Multiple-mappings).

**Comment by ducman** [comment]
So how would you do something like the below:
```sql
UPDATE User SET IsActive = 1 WHERE Id IN (1,2,3,4)
-- or
UPDATE User SET IsActive = 0 WHERE EmailAddress LIKE '%@gmail.com'
````

**Comment by MoonStorm** [comment]
Wouldn't an entity mapping override that only covers ``IsActive`` together with a WHERE clause work for this case?

**Comment by ducman** [comment]
Hmmm, I guess you're right. I can't think of any other scenario since you can accomplish it by building out the entire where clause. 

The only thing I would say is to remove the ability to add parameters to the `statementOptions` for just bulk update so it won't cause confusing.

**Comment by MoonStorm** [comment]
Don't understand your statement. The parameters  are usually used for the where clause.

**Comment by ducman** [comment]
Consider the below query:
```sql
UPDATE Business SET Name = @Name WHERE Name LIKE @keyword + '%'
```

The dapper/fastcrud equivalent would be:
```c#
public void UpdateNameForAll(string newName, string keyword)
{
  var partialUpdateMapping = OrmConfiguration
    .GetDefaultEntityMapping<Business>()
    .Clone()
    .RemoveAllPropertiesExcluding(nameof(Business.Name));
 
  databaseConnection.BulkUpdate(new Business { Name = newName}, statement => statement
    .Where($"{nameof(Business.Name):C} LIKE @keyword + '%'")
    .WithParameters(new { keyword })
    .WithEntityMappingOVerride(partialUpdateMapping));
}
```
The above throws an exception because `@keyword` hasn't been declared. The `WithParameters(...)` IS NOT used in the a bulk update **but you can still call it**.

Because we can't use parameters, the solution would be this:
```c#
public void UpdateNameForAll(string newName, string keyword)
{
  var partialUpdateMapping = OrmConfiguration
    .GetDefaultEntityMapping<Business>()
    .Clone()
    .RemoveAllPropertiesExcluding(nameof(Business.Name));
 
  databaseConnection.BulkUpdate(new Business { Name = newName}, statement => statement
    .Where($"{nameof(Business.Name):C} LIKE '{keyword}%'")
    // .WithParameters(new { keyword }) no need for this since it's not used
    .WithEntityMappingOVerride(partialUpdateMapping));
}
```

Though this open it up to sql injection (different topic and probably why I prefer parameters). But what I'm trying to get at is that the `WithParameters(...)` should be removed since it will confuse other developers since it is available, yet it's not used at all in a bulk update. And I'm talking about just bulk update. I hope that make sense.

**Comment by MoonStorm** [comment]
Yes that does make sense. If ``WithParameters`` is not used in the bulk update, that is obviously an  oversight.

**Comment by fuzzzerd** [comment]
I too have run across this issue. Would be nice to either have the call to `.WithParameters(..)` work as expected and/outlined by @ducman or at the minimum throw a `NotSupportedException` when called for Bulk Updates.

I took a browse through the code and it looks like the issue stems from the `.Execute()` extension. In the GenericSqlStatements class, on the bulk update methods [the entity itself is passed in as the parameter, ](https://github.com/MoonStorm/Dapper.FastCRUD/blob/master/Dapper.FastCrud/SqlStatements/GenericSqlStatements.cs#L203) instead of [`statementOptions.Parameters` as it is in the other cases such as Bulk Delete](https://github.com/MoonStorm/Dapper.FastCRUD/blob/master/Dapper.FastCrud/SqlStatements/GenericSqlStatements.cs#L242).

**Comment by dstringvc** [comment]
2 1/2 years after the original post, this is still confusing developers. @ducman , thanks for the work around!

**Comment by baleeds** [comment]
@MoonStorm are there plans to change this behavior in the bulk update?

**Comment by MoonStorm** [comment]
The problem in this case is that both the entity with the updated properties AND the extra parameters both need to be passed in the query as a single parameter object. But this can be solved with **DynamicProperties**.
